drivers/eeprom/eeprom_emulator.c

```diff
@@ -0,0 +1,748 @@
+				if (rc) {
+					return rc;
+				}
+
+			}
+
+			rd_offset += sizeof(buf);
+		}
+
+	}
+
+	/* invalidate the old page */
+	rc = eeprom_emu_page_invalidate(dev, dev_data->page_offset);
+	if (!rc) {
+		dev_data->write_offset = dev_config->size;
+		dev_data->page_offset = next_page_offset;
+	}
+
+	return rc;
+}
+
+
+/*
+ * eeprom_emu_word_write: write basic word (cbs bytes of data) item to address,
+ */
+static int eeprom_emu_word_write(const struct device *dev, off_t address,
+				 const uint8_t *data)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	struct eeprom_emu_data *dev_data = DEV_DATA(dev);
+	uint8_t buf[dev_config->flash_cbs], tmp[dev_config->flash_cbs];
+	off_t direct_address, wraddr;
+	int rc;
+
+	direct_address = dev_data->page_offset + address;
+
+	rc = eeprom_emu_flash_read(dev, direct_address, buf, sizeof(buf));
+	if (rc) {
+		return rc;
+	}
+
+	if (!eeprom_emu_is_word_used(dev, buf)) {
+		if (eeprom_emu_is_word_used(dev, data)) {
+			rc = eeprom_emu_flash_write(dev, direct_address, data,
+						    sizeof(buf));
+		}
+
+		goto end;
+	}
+
+	rc = eeprom_emu_word_read(dev, address, buf);
+	if (rc) {
+		return rc;
+	}
+
+	if (!memcmp(buf, data, sizeof(buf))) {
+		/* data has not changed */
+		goto end;
+	}
+
+	wraddr = address;
+	/* store change */
+	for (uint8_t i = 0; i < 2; i++) {
+		if (memcmp(&buf[i*sizeof(buf)/2], data, sizeof(buf)/2)) {
+			eeprom_emu_set_change(dev, wraddr, data, tmp);
+			rc = eeprom_emu_flash_write(dev, dev_data->page_offset +
+						    dev_data->write_offset, tmp,
+						    sizeof(buf));
+			if (rc) {
+				return rc;
+			}
+
+			dev_data->write_offset += sizeof(buf);
+			if ((dev_data->write_offset + sizeof(buf)) >=
+			    dev_config->page_size) {
+				rc = eeprom_emu_compactor(dev);
+				if (rc) {
+					return rc;
+				}
+
+			}
+
+		}
+
+		data += sizeof(buf)/2;
+		wraddr += sizeof(buf)/2;
+	}
+
+	/* reset data pointer to start */
+	data -= sizeof(buf);
+end:
+	if (dev_config->rambuf) {
+		memcpy(dev_config->rambuf + address, data, sizeof(buf));
+	}
+
+	return rc;
+}
+
+static int eeprom_emu_range_is_valid(const struct device *dev, off_t address,
+				     size_t len)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+
+	if ((address + len) <= dev_config->size) {
+		return 1;
+	}
+
+	return 0;
+}
+
+/*
+ * eeprom_emu_rw: read and write routine
+ */
+static int eeprom_emu_rw(const struct device *dev, off_t address, void *data,
+			 size_t len, bool write_op)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	struct eeprom_emu_data *dev_data = DEV_DATA(dev);
+	uint8_t *data8 = (uint8_t *)data;
+	uint8_t buf[dev_config->flash_cbs];
+	off_t addr_jmp;
+	int rc = 0;
+
+	if ((!data) || (!eeprom_emu_range_is_valid(dev, address, len))) {
+		return -EINVAL;
+	}
+
+	if (!len) {
+		return 0;
+	}
+
+	if (!device_is_ready(dev_config->flash_dev)) {
+		LOG_ERR("flash device is not ready");
+		return -EIO;
+	}
+
+	if ((write_op) && (dev_config->readonly)) {
```

Comment:
Maybe switch the `if (!device` with this one in places; this basically serves bad parameters to the function and there is no need for checking if device is ready if function parameters will fail further operation.

Reply:
I will recheck the order and definitively move the return for len=0 down.

Reply:
After reconsidering I have changed the order:
1. Check if there is nothing to do and return early
2. Do input error checking and return if error
3. Handle normal case

I will try to keep this in next pieces of code :-) :-).